internal/datasystem/fdv2_datasystem.go

```diff
@@ -205,35 +222,99 @@ func (f *FDv2) runInitializers(ctx context.Context, closeWhenReady chan struct{}
+// closeWhenReady, and it uses a sync.Once to ensure that happens only once.
+//
+// This design could be improved significantly. Some ideas:
+// 1) Refactor the SDK to listen to Data Source status rather than the special closeWhenReady channel. It's unclear why
+// this isn't currently the case, but there may be good reasons.
+// 2) Use callbacks. Somewhat equivalent to (1), but instead of needing to wrap the channel, we just proxy directly to
+// calling the sync.Once.
+// 3) Make the channel multi-use. Instead of the contract being "close when ready", have it be "send a value when ready".
+func (f *FDv2) closeChannelWrapper(ctx context.Context, closeWhenReady chan struct{}) chan struct{} {
+	ch := make(chan struct{})
+	go func() {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ch:
+			f.readyOnce.Do(func() {
+				close(closeWhenReady)
+			})
+		}
+	}()
+	return ch
+}
+
+func (f *FDv2) runSynchronizers(ctx context.Context, closeWhenReady chan struct{}) {
+	// If the SDK was configured with no synchronizer, then (assuming no initializer succeeded, which would have
+	// already closed the channel), we should close it now so that MakeClient unblocks.
 	if f.primarySync == nil {
 		f.readyOnce.Do(func() {
 			close(closeWhenReady)
 		})
 		return
 	}
 
-	// We can't pass closeWhenReady to the data source, because it might have already been closed.
-	// Instead, create a "proxy" channel just for the data source; if that is closed, we close the real one
-	// using the sync.Once.
-	ready := make(chan struct{})
-	f.primarySync.Sync(ready, selector)
+	f.launchTask(func() {
+		for {
+			f.loggers.Debugf("Primary synchronizer %s is starting", f.primarySync.Name())
+			f.primarySync.Sync(f.closeChannelWrapper(ctx, closeWhenReady), f.store.Selector())
+			if err := f.evaluateCond(ctx, f.fallbackCond); errors.Is(err, context.Canceled) {
+				return
+			}
+			if err := f.primarySync.Close(); err != nil {
+				f.loggers.Errorf("Primary synchronizer %s failed to gracefully close: %v", f.primarySync.Name(), err)
+			}
+			f.loggers.Debugf("Fallback condition met")
+			f.loggers.Debugf("Secondary synchronizer %s is starting", f.secondarySync.Name())
 
+			f.UpdateStatus(interfaces.DataSourceStateInterrupted, interfaces.DataSourceErrorInfo{})
+
+			f.secondarySync.Sync(f.closeChannelWrapper(ctx, closeWhenReady), f.store.Selector())
+			if err := f.evaluateCond(ctx, f.recoveryCond); errors.Is(err, context.Canceled) {
+				return
+			}
+			if err := f.secondarySync.Close(); err != nil {
+				f.loggers.Errorf("Secondary synchronizer %s failed to gracefully close: %v", f.secondarySync.Name(), err)
+			}
+			f.loggers.Debugf("Recovery condition met")
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+		}
+	})
+}
+
+func (f *FDv2) evaluateCond(ctx context.Context, cond func(status interfaces.DataSourceStatus) bool) error {
+	ticker := time.NewTicker(10 * time.Second)
```

Comment:
This hard coded 10 second timer is what limits the resolution of the fallback / recovery conditions right?

Reply:
Correct. I could see an alternative of making this event-triggered, where we'd have "timer event" and "data source status event" (and anything else that can be used as a condition.)

But then we'd need to hold a map of timers, hook into the data source status broadcasters.. it just doesn't seem worth the complexity when compared to a predictable "tick".